### ISSUE TITLE
Haze removal doesn't work with images with large overexposed regions 

### ISSUE THREAD

**Issue by chaimav** [Beep6581/RawTherapee#5181]
When editing an image with a large overexposed area, the image preview of the Haze Removal tool  does not show changes, but changes do show up in the in the thumbnail.
Thread about issue with representative raw file here: https://discuss.pixls.us/t/haze-removal-bug/11423

RawTherapee info:
Version: 5.5
Branch: 5.5
Commit: d9ac23555
Commit date: 2018-12-17
Compiler: gcc 8.2.1
Processor: generic x86
System: Windows
Bit depth: 64 bits
Gtkmm: V3.22.3
Lensfun: V0.3.2.0
Build type: release
Build flags:  -m64 -mwin32 -msse2 -mfpmath=sse -mthreads -Wno-aggressive-loop-optimizations -Wno-parentheses -std=c++11 -mtune=generic -Werror=unused-label -Wall -Wuninitialized -Wno-deprecated-declarations -Wno-unused-result -fopenmp -Werror=unknown-pragmas  -DNDEBUG  -O3 -ftree-vectorize
Link flags: -m64  -mthreads  -static-libgcc   -mtune=generic  -s  -O3  -fno-use-linker-plugin
OpenMP support: ON
MMAP support: ON

Windows 10

**Comment by agriggio** [comment]
This should do (it also incorporates a fix for a race condition, thanks @heckflosse)

```diff
diff --git a/rtengine/ipdehaze.cc b/rtengine/ipdehaze.cc
--- a/rtengine/ipdehaze.cc
+++ b/rtengine/ipdehaze.cc
@@ -63,14 +63,12 @@
     const int W = R.width();
     const int H = R.height();
 
-    int npatches = 0;
-
 #ifdef _OPENMP
     #pragma omp parallel for if (multithread)
 #endif
     for (int y = 0; y < H; y += patchsize) {
         int pH = min(y+patchsize, H);
-        for (int x = 0; x < W; x += patchsize, ++npatches) {
+        for (int x = 0; x < W; x += patchsize) {
             float val = RT_INFINITY_F;
             int pW = min(x+patchsize, W);
             for (int yy = y; yy < pH; ++yy) {
@@ -97,7 +95,7 @@
         }
     }
 
-    return npatches;
+    return (W / patchsize + ((W % patchsize) > 0)) * (H / patchsize + ((H % patchsize) > 0));
 }
 
 
@@ -121,7 +119,7 @@
         std::priority_queue<float> p;
         for (int y = 0; y < H; y += patchsize) {
             for (int x = 0; x < W; x += patchsize) {
-                if (!OOG(dark[y][x], 1.f)) {
+                if (!OOG(dark[y][x], 1.f - 1e-5f)) {
                     p.push(dark[y][x]);
                 }
             }
```

**Comment by heckflosse** [comment]
@agriggio The patch does not apply to dev because I already pushed another fix for the race (you know) with e2154f9.
Nevertheless we should apply both changes to dev code. Shall I do or will you do?

**Comment by agriggio** [comment]
@heckflosse ah sorry, I didn't notice :-(
Please go ahead if you can, thanks!
